cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginService.java

```diff
@@ -0,0 +1,221 @@
+      // should not happen
+      throw new ArtifactNotFoundException(artifact);
+    }
+
+    // return the first one from the artifact details list.
+    return artifactDetails.get(0).getDescriptor();
+  }
+
+  private PluginEndpoint getPluginEndpoint(Id.Namespace namespace, ArtifactDetail artifactDetail, String pluginType,
+                                           String pluginName, ArtifactDescriptor parentArtifactDescriptor,
+                                           String methodName)
+    throws NotFoundException, IOException, ClassNotFoundException {
+
+    Id.Artifact artifactId = Id.Artifact.from(namespace, artifactDetail.getDescriptor().getArtifactId());
+    PluginClass pluginClass =
+      artifactRepository.findPlugin(namespace, artifactId, pluginType, pluginName, new PluginSelector()).getValue();
+
+    // found plugin, check for endpoint
+    if (pluginClass.getEndpoints() == null || !pluginClass.getEndpoints().contains(methodName)) {
+      throw new NotFoundException(
+        String.format("Plugin with type: %s name: %s found, " +
+                        "but Endpoint %s was not found", pluginType, pluginName, methodName));
+    }
+
+    // initialize parent classloader and plugin instantiator
+    parentClassLoader = artifactClassLoaderFactory.createClassLoader(parentArtifactDescriptor.getLocation());
+    pluginInstantiator = new PluginInstantiator(cConf, parentClassLoader, stageDir);
+
+    // add plugin artifact
+    pluginInstantiator.addArtifact(artifactDetail.getDescriptor().getLocation(), artifactId.toArtifactId());
+
+    // we pass the parent artifact to endpoint plugin context,
+    // as plugin method will use this context to load other plugins.
+    DefaultEndpointPluginContext defaultEndpointPluginContext =
+      new DefaultEndpointPluginContext(namespace.toEntityId(), artifactRepository, pluginInstantiator,
+                                       Id.Artifact.from(namespace, parentArtifactDescriptor.getArtifactId()));
+
+    return getPluginEndpoint(pluginInstantiator, artifactId,
+                             pluginClass, methodName, defaultEndpointPluginContext);
+  }
+
+  /**
+   * load and instantiate the plugin and return {@link PluginEndpoint}
+   * which can be used to invoke plugin method with request object.
+   *
+   * @param pluginInstantiator to instantiate plugin instances.
+   * @param artifact artifact of the plugin
+   * @param pluginClass class having the plugin method to invoke
+   * @param endpointName name of the endpoint to invoke
+   * @param endpointPluginContext endpoint plugin context that can optionally be passed to the method
+   * @throws IOException if there was an exception getting the classloader
+   * @throws ClassNotFoundException if plugin class cannot be loaded
+   * @throws NotFoundException Not Found exception thrown if no matching plugin found.
+   */
+  public PluginEndpoint getPluginEndpoint(final PluginInstantiator pluginInstantiator, Id.Artifact artifact,
```

Comment:
there are two methods to get a PluginEndpoint?  When do you use one versus the other?